source/dub/dependencyresolver.d

```diff
@@ -197,8 +197,16 @@ class DependencyResolver(CONFIGS, CONFIG) {
 		}
 
 		string first_error;
+		size_t loop_counter = 0;
 
 		while (true) {
+			assert(loop_counter++ < 1_000_000,
+				"The dependency resolution process is taking too long. The"
+				~ " dependency graph is likely hitting a pthological case in"
```

Comment:
Typo: "pathological".

Also I *feel* there should be a way to bypass this. I would prefer to issue a warning at first instead of crashing the process, but since it is aimed at stuck CI, maybe just add an option to bypass this behavior so someone with such a pathological case still has a working project.

Reply:
The problem is that when reaching this amount of iterations, it is likely that the total number lies in the hundreds of millions or billions and the process will not terminate in any reasonable amount of time. At least that is the experience so far, either it terminates after a few hundred iterations, or "never". The reason for this also lies in the simple lexicographical order in which combinations are tested

Reply:
I can understand why `assert` by default is prefered, but I don't think we should ever allow a false positive to happen without a workaround.
If there's one single user which have a project which is working currently, that this change will break, it's not acceptable to pull this change, because for this person *you're not improving things*.

My suggestion would be to use an hidden switch here. Now if you feel strongly about it, I won't push it, 1M sound already quite high.
I just saw users' use case being overlooked too many times because "why would you do that?".

Reply:
Okay, just based on gut feeling I think the likelihood for a false positive is very low here, and even if there is one, nobody would probably rely on it, considering a DUB invocation would take more than an hour. But on the other hand it's not really a big deal to add a way to disable the assert. Instead of a switch I'd suggest checking for a `DUB_NO_RESOLVE_LIMIT` environment variable, because that only has local code impact and may also be easier to incorporate in situations where DUB is invoked many times through a script.